tokens/service.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2021 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package tokens
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storjscan/tokens/erc20"
+)
+
+// ErrService - tokens service error class.
+var ErrService = errs.Class("tokens service")
+
+// Config holds tokens service configuration.
+type Config struct {
+	Endpoint     string
+	TokenAddress string
+}
+
+// Service for querying ERC20 token information from ethereum chain.
+//
+// architecture: Service
+type Service struct {
+	log      *zap.Logger
+	endpoint string
+	token    Address
+}
+
+// NewService creates new token service instance.
+func NewService(log *zap.Logger, endpoint string, token Address) *Service {
+	return &Service{
+		log:      log,
+		endpoint: endpoint,
+		token:    token,
+	}
+}
+
+// Payments retrieves all ERC20 token payments for ethereum address.
+func (service *Service) Payments(ctx context.Context, address Address) (_ []Payment, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	client, err := ethclient.DialContext(ctx, service.endpoint)
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	token, err := erc20.NewErc20(service.token, client)
+	if err != nil {
+		return nil, ErrService.Wrap(err)
+	}
+
+	opts := &bind.FilterOpts{
+		Start:   0,
+		End:     nil,
+		Context: ctx,
+	}
+	iter, err := token.FilterTransfer(opts, nil, []Address{address})
```

Comment:
One more generic comment: I have problems earlier when I tried to query too may record. For example, infura fails with `Error: Couldn't get logs for contract 0xB64ef51C888972c908CFacf59B47C1AfBC0Ab8aC: query returned more than 10000 results`.

Is it possible here to get too many records? (or get it too slowly?) Or is a full node returns with all the records?

Reply:
So that PR is more about setting basic project infrastructure and showing how service is going to interact with a node to get necessary ERC20 token information and how it is going to be tested. So this payments API implementation is rather naive at this point and just a MVP.